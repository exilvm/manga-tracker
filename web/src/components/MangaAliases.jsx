import { makeStyles } from '@material-ui/core/styles';
import { DoubleArrow as DoubleArrowIcon } from '@material-ui/icons';

import React, { useCallback } from 'react';
import {
  IconButton,
  List,
  ListItem,
  ListItemText,
  Typography, Tooltip,
} from '@material-ui/core';

import PropTypes from 'prop-types';
import { useUser } from '../utils/useUser';

const useStyles = makeStyles((theme) => ({
  root: {
    margin: theme.spacing(1),
    maxHeight: '250px',
    width: 'max-content',
  },
  list: {
    overflow: 'auto',
    maxHeight: '100%',
  },
  listItem: {
    marginTop: '0px',
    marginBottom: '0px',
  },
  promoteIcon: {
    transform: 'rotate(-90deg)',
  },
}));

const MangaAliases = (props) => {
  const {
    aliases,
    mangaId,
    onTitleUpdate,
    enqueueSnackbar,
    confirm,
    allowEdits=false,
  } = props;

  const classes = useStyles();
  const { isAdmin } = useUser();
  const autoHideDuration = 8000;

  const onAliasPromote = useCallback((title) => {
    confirm({
      description: `Do you want to set "${title}" as the main title for this manga?`,
      confirmationText: 'Yes',
      cancellationText: 'No',
    }).then(() => {
      fetch(`/api/admin/manga/${mangaId}/title`, {
        method: 'POST',
        credentials: 'same-origin',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify({ title }),
      })

        .then(res => res.json())

        .then(json => {
          if (json.error) {
            enqueueSnackbar(
              `Failed to set alias as main title. ${json.error}`,
              { variant: 'error', autoHideDuration }
            );
          } else if (onTitleUpdate) {
            enqueueSnackbar(
              `Set "${title}" as the main title. ${json.message}`,
              { variant: 'success', autoHideDuration }
            );
            onTitleUpdate();
          }
        })

        .catch(err => enqueueSnackbar(
          `Failed to set title. ${err}`,
          { variant: 'error', autoHideDuration }
        ));
    })
      .catch(() => {});
  }, [enqueueSnackbar, mangaId, onTitleUpdate, confirm]);

  if (!aliases || aliases.length === 0) return null;

  return (
    <div id='manga-aliases' className={classes.root}>
      <Typography>Alternative titles</Typography>
      <List className={classes.list}>
        {aliases.map(alias => (
          <ListItem key={alias} dense>
            <ListItemText primary={alias} className={classes.listItem} />
            {allowEdits && isAdmin && (
              <Tooltip title='Set as main title'>
                <IconButton
                  size='small'
                  onClick={() => onAliasPromote(alias)}
                  aria-label='Set alias as main title'
                >
                  <DoubleArrowIcon className={classes.promoteIcon} />
                </IconButton>
              </Tooltip>
            )}
          </ListItem>
        ))}
      </List>
    </div>
  );
};

MangaAliases.propTypes = {
  aliases: PropTypes.arrayOf(PropTypes.string),
  mangaId: PropTypes.number,
  onTitleUpdate: PropTypes.func,
  enqueueSnackbar: PropTypes.func,
  confirm: PropTypes.func,
  allowEdits: PropTypes.bool,
};

export default MangaAliases;
