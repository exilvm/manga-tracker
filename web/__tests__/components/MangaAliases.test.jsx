import React from 'react';
import { render, screen, act, within } from '@testing-library/react';
import { create } from 'react-test-renderer';

import MangaAliases from '../../src/components/MangaAliases';

describe('MangaAliases renders correctly', () => {
  it('Should render correctly without aliases', () => {
    expect(create(<MangaAliases />)).toMatchSnapshot();
    expect(create(<MangaAliases aliases={[]} />)).toMatchSnapshot();
  });

  it('Should render correctly with aliases', () => {
    expect(create(<MangaAliases aliases={['a', 'b', 'c']} />)).toMatchSnapshot();
  });

  it('Should render correctly with editing allowed', () => {
    expect(create(<MangaAliases aliases={['a', 'b', 'c']} allowEdits />)).toMatchSnapshot();
  });
});
