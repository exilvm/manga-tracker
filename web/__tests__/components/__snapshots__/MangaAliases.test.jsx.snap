// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MangaAliases renders correctly Should render correctly with aliases 1`] = `
<div
  className="makeStyles-root-1"
  id="manga-aliases"
>
  <p
    className="MuiTypography-root MuiTypography-body1"
  >
    Alternative titles
  </p>
  <ul
    className="MuiList-root makeStyles-list-2 MuiList-padding"
  >
    <li
      className="MuiListItem-root MuiListItem-dense MuiListItem-gutters"
      disabled={false}
    >
      <div
        className="MuiListItemText-root makeStyles-listItem-3 MuiListItemText-dense"
      >
        <span
          className="MuiTypography-root MuiListItemText-primary MuiTypography-body2 MuiTypography-displayBlock"
        >
          a
        </span>
      </div>
    </li>
    <li
      className="MuiListItem-root MuiListItem-dense MuiListItem-gutters"
      disabled={false}
    >
      <div
        className="MuiListItemText-root makeStyles-listItem-3 MuiListItemText-dense"
      >
        <span
          className="MuiTypography-root MuiListItemText-primary MuiTypography-body2 MuiTypography-displayBlock"
        >
          b
        </span>
      </div>
    </li>
    <li
      className="MuiListItem-root MuiListItem-dense MuiListItem-gutters"
      disabled={false}
    >
      <div
        className="MuiListItemText-root makeStyles-listItem-3 MuiListItemText-dense"
      >
        <span
          className="MuiTypography-root MuiListItemText-primary MuiTypography-body2 MuiTypography-displayBlock"
        >
          c
        </span>
      </div>
    </li>
  </ul>
</div>
`;

exports[`MangaAliases renders correctly Should render correctly with editing allowed 1`] = `
<div
  className="makeStyles-root-1"
  id="manga-aliases"
>
  <p
    className="MuiTypography-root MuiTypography-body1"
  >
    Alternative titles
  </p>
  <ul
    className="MuiList-root makeStyles-list-2 MuiList-padding"
  >
    <li
      className="MuiListItem-root MuiListItem-dense MuiListItem-gutters"
      disabled={false}
    >
      <div
        className="MuiListItemText-root makeStyles-listItem-3 MuiListItemText-dense"
      >
        <span
          className="MuiTypography-root MuiListItemText-primary MuiTypography-body2 MuiTypography-displayBlock"
        >
          a
        </span>
      </div>
    </li>
    <li
      className="MuiListItem-root MuiListItem-dense MuiListItem-gutters"
      disabled={false}
    >
      <div
        className="MuiListItemText-root makeStyles-listItem-3 MuiListItemText-dense"
      >
        <span
          className="MuiTypography-root MuiListItemText-primary MuiTypography-body2 MuiTypography-displayBlock"
        >
          b
        </span>
      </div>
    </li>
    <li
      className="MuiListItem-root MuiListItem-dense MuiListItem-gutters"
      disabled={false}
    >
      <div
        className="MuiListItemText-root makeStyles-listItem-3 MuiListItemText-dense"
      >
        <span
          className="MuiTypography-root MuiListItemText-primary MuiTypography-body2 MuiTypography-displayBlock"
        >
          c
        </span>
      </div>
    </li>
  </ul>
</div>
`;

exports[`MangaAliases renders correctly Should render correctly without aliases 1`] = `null`;

exports[`MangaAliases renders correctly Should render correctly without aliases 2`] = `null`;
